### PR TITLE
Make tokenListener plural

### PR DESCRIPTION
**Pull request by chuckbergeron**
* this changes pool.tokens.tokenListener to pool.tokens.tokenListeners as an array of token faucets and drip tokens

### TokenFaucets:
Screenshot 2021-07-08 at 13 31 37

### Tokens list (necessary to collect USD / derivedETH data):
Screenshot 2021-07-08 at 13 31 17

### CODE REVIEW

File: packages/api-runner/lib/fetchers/getPoolChainData.js [hunk 1]
```diff
@@ -89,21 +91,11 @@ export const getPoolChainData = async (chainId, poolGraphData, fetch) => {
...
-      )
-      batchCalls.push(
-        tokenFaucetContract
-          .dripRatePerSecond()
-          .asset()
-          .measure()
-      )
-    }
+    // Token Faucets
+    pool.tokenFaucets.forEach((tokenFaucetAddress) => {
```
Comment (dylandesrosier) [inline packages/api-runner/lib/fetchers/getPoolChainData.js]:
```suggestion
    pool.tokenFaucets?.forEach((tokenFaucetAddress) => {
```

So we're not dependent on adding the empty array to the `current-pool-data`

File: packages/api-runner/lib/fetchers/getPoolChainData.js [hunk 2]
```diff
@@ -218,21 +206,21 @@ export const getPoolChainData = async (chainId, poolGraphData, fetch) => {
...
-      )
-      batchCalls.push(
-        dripErc20Contract
-          .balanceOf(pool.tokenListener.address)
-          .decimals()
-          .symbol()
-          .name()
-      )
-    }
+    pool.tokenFaucets.forEach((tokenFaucetAddress) => {
```
Comment (dylandesrosier) [inline packages/api-runner/lib/fetchers/getPoolChainData.js]:
```suggestion
    pool.tokenFaucets?.forEach((tokenFaucetAddress) => {
```

Same thing here

File: packages/api-runner/lib/fetchers/getPoolChainData.js [hunk 3]
```diff
@@ -409,43 +392,56 @@ const formatPoolChainData = (
...
-      const tokenFaucetDripToken = {
-        address: tokenFaucetDripAssetAddress.toLowerCase(),
-        amount: formatUnits(tokenListenerData2.balanceOf[0], tokenListenerData2.decimals[0]),
-        amountUnformatted: tokenListenerData2.balanceOf[0],
-        decimals: tokenListenerData2.decimals[0],
-        name: tokenListenerData2.name[0],
-        symbol: tokenListenerData2.symbol[0]
+
+      const dripToken = {
+        address: dripTokenAddress.toLowerCase(),
```
Comment (dylandesrosier) [inline packages/api-runner/lib/fetchers/getPoolChainData.js:406]: Could we add some id so you can link the drip token from this array back to the specific token faucet. Just `tokenFaucetAddress: tokenFaucet.address` would be enough.

File: packages/api-runner/lib/fetchers/getPools.js [hunk 4]
```diff
@@ -163,22 +170,39 @@ const combineTokenPricesData = (_pools, tokenPriceData, defaultTokenPriceUsd) =>
...
+      addTokenTotalUsdValue(tokenFaucet.dripToken, tokenPriceData, defaultTokenPriceUsd)
+    })
+
     // Add to all known tokens
-    Object.values(pool.tokens).forEach((token) =>
-      addTokenTotalUsdValue(token, tokenPriceData, defaultTokenPriceUsd)
-    )
+    Object.values(pool.tokens).forEach((token) => {
+      // This takes care of tokenFaucetDripTokens:
+      if (token?.length > 0) {
```
Comment (dylandesrosier) [inline packages/api-runner/lib/fetchers/getPools.js]:
```suggestion
      if (Array.isArray(token)) {
```

Just a little clearer

File: packages/api-runner/lib/fetchers/getPools.js [hunk 5]
```diff
@@ -530,19 +557,23 @@ const calculateTotalValueLockedPerPool = (pools) =>
  * @param {*} pools
  * @returns
  */
-const calculateTokenFaucetApr = (pools) =>
+const calculateTokenFaucetAprs = (pools) =>
```
Comment (dylandesrosier) [inline packages/api-runner/lib/fetchers/getPools.js:562]: Should we add this to `tokens.tokenFaucetDripTokens` as well?

The more I think about it the more I think having them accessible from `tokenFaucets[0].dripToken` is enough and that we should just drop `tokens.tokenFaucetDripTokens`.

**Review by dylandesrosier — APPROVED** [review]

File: packages/api-runner/lib/fetchers/getPools.js (hunk 5 above)
Comment (chuckbergeron) [reply packages/api-runner/lib/fetchers/getPools.js:562]: 1) No, as the APR can differ from faucet to faucet. It's a faucet-specific thing.

2) We need to have the drip token assets in the `tokens` list in order to get token price data. So we _could_ clear that key out from `tokens` after the fact, but I felt it made more sense to simply leave it there.

**Pull request merged**